src/auth.ts

```diff
@@ -0,0 +1,283 @@
+import {
+  app,
+  BrowserWindow,
+  ipcMain,
+  IpcMainEvent,
+  safeStorage,
+  shell,
+} from 'electron'
+import { Profile, ProfileSchema, UserInfo } from './schemas/profile'
+import { authenticate } from './services/grafana/authenticate'
+import path from 'path'
+import { writeFile, readFile, rm } from 'fs/promises'
+import { fetchInstances } from './services/grafana'
+import { fetchPersonalToken } from './services/k6'
+import { SignInProcessState, SignInResult, Stack } from './types/auth'
+import log from 'electron-log/main'
+import { ClientError } from 'openid-client'
+
+interface WaitForDone<T> {
+  status: 'done'
+  data: T
+}
+
+interface WaitForAborted {
+  status: 'aborted'
+}
+
+type WaitForResult<T> = WaitForDone<T> | WaitForAborted
+
+interface WaitForOptions {
+  event: string
+  signal: AbortSignal
+  timeout?: number
+}
+
+async function waitFor<T>({ event, signal, timeout }: WaitForOptions) {
+  return new Promise<WaitForResult<T>>((resolve, reject) => {
+    const timeoutId =
+      timeout &&
+      setTimeout(() => {
+        ipcMain.removeListener(event, handleMessage)
+
+        reject(new Error(`Timeout waiting for "${event}"`))
+      }, timeout)
+
+    const handleAbort = () => {
+      clearTimeout(timeoutId)
+
+      ipcMain.removeListener(event, handleMessage)
+
+      resolve({
+        status: 'aborted',
+      })
+    }
+
+    const handleMessage = (_: IpcMainEvent, data: T) => {
+      clearTimeout(timeoutId)
+
+      signal.removeEventListener('abort', handleAbort)
+
+      resolve({
+        status: 'done',
+        data,
+      })
+    }
+
+    signal.addEventListener('abort', handleAbort)
+
+    ipcMain.once(event, handleMessage)
+  })
+}
+
+const fileName =
+  process.env.NODE_ENV === 'development'
+    ? 'k6-studio-profile-dev.json'
+    : 'k6-studio-profile.json'
+
+const filePath = path.join(app.getPath('userData'), fileName)
+
+function wasAborted(error: unknown): boolean {
+  return (
+    (error instanceof ClientError && error.code == 'OAUTH_ABORT') ||
+    (error instanceof Error && error.name === 'AbortError')
+  )
+}
+
+async function getCurrentProfile() {
+  try {
+    const file = await readFile(filePath, 'utf-8')
+
+    return ProfileSchema.parse(JSON.parse(file))
+  } catch {
+    return null
+  }
+}
+
+export function initAuth(browserWindow: BrowserWindow) {
+  function notifyStateChange(state: SignInProcessState) {
+    browserWindow.webContents.send('auth:state-change', state)
+  }
+
+  ipcMain.handle('auth:get-user', async (): Promise<UserInfo | null> => {
+    const profile = await getCurrentProfile()
+
+    if (profile === null) {
+      return null
+    }
+
+    return profile.user
+  })
+
+  let pending: AbortController | null = null
+
+  ipcMain.handle('auth:sign-in', async (): Promise<SignInResult> => {
```

Comment:
This function basically acts like a co-routine, where it sends the latest state to the UI and waits for the user doing something, e.g. selecting a stack. Once the entire process has completed, the result of the sign in is returned.